quinn-proto/src/connection/timer.rs

```diff
@@ -1,123 +1,356 @@
+    PathAbandoned = 7,
     /// When the peer fails to confirm abandoning the path
-    PathNotAbandoned(PathId),
+    PathNotAbandoned = 8,
+}
+
+impl PerPathTimer {
+    const VALUES: [Self; 9] = [
+        Self::LossDetection,
+        Self::PathIdle,
+        Self::PathValidation,
+        Self::PathOpen,
+        Self::PathKeepAlive,
+        Self::Pacing,
+        Self::MaxAckDelay,
+        Self::PathAbandoned,
+        Self::PathNotAbandoned,
+    ];
 }
 
 /// Keeps track of the nearest timeout for each `Timer`
 ///
 /// The [`TimerTable`] is advanced with [`TimerTable::expire_before`].
 #[derive(Debug, Clone, Default)]
 pub(crate) struct TimerTable {
-    most_recent_timeout: FxHashMap<Timer, Instant>,
-    timeout_queue: BinaryHeap<TimerEntry>,
+    generic: [Option<Instant>; GenericTimer::VALUES.len()],
+    path_timers: SmallMap<PathId, PathTimerTable, STACK_TIMERS>,
+}
+
+/// For how many paths we keep the timers on the stack, before spilling onto the heap.
+const STACK_TIMERS: usize = 4;
+
+/// Works like a `HashMap` but stores up to `SIZE` items on the stack.
+#[derive(Debug, Clone)]
+struct SmallMap<K, V, const SIZE: usize> {
+    stack: [Option<(K, V)>; SIZE],
+    heap: Option<IntMap<K, V>>,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub(super) struct TimerEntry {
-    pub(super) time: Instant,
-    pub(super) timer: Timer,
+impl<K, V, const SIZE: usize> Default for SmallMap<K, V, SIZE> {
+    fn default() -> Self {
+        Self {
+            stack: [const { None }; SIZE],
+            heap: None,
+        }
+    }
 }
 
-impl Ord for TimerEntry {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        // `timeout_queue` is a max heap so we need to reverse the order to efficiently pop the
-        // next timeout
-        self.time
-            .cmp(&other.time)
-            .then_with(|| self.timer.cmp(&other.timer))
-            .reverse()
+impl<K, V, const SIZE: usize> SmallMap<K, V, SIZE>
+where
+    K: std::cmp::Eq + std::hash::Hash + IdentityHashable,
+{
+    fn insert(&mut self, key: K, value: V) -> Option<V> {
+        // check stack for space
+        for el in self.stack.iter_mut() {
+            match el {
+                Some((k, v)) => {
+                    if *k == key {
+                        let old_value = std::mem::replace(v, value);
+                        return Some(old_value);
+                    }
+                }
+                None => {
+                    // make sure to remove a potentially old value from the heap
+                    let old_heap = self.heap.as_mut().and_then(|h| h.remove(&key));
+                    *el = Some((key, value));
+
+                    return old_heap;
+                }
+            }
+        }
+
+        // No space on the stack, use the heap
+        if self.heap.is_none() {
+            self.heap = Some(Default::default());
+        }
+        let heap = self.heap.as_mut().expect("inserted");
+        heap.insert(key, value)
```

Comment:
`self.heap.get_or_insert_default().insert(key, value);` would do this without `.expect()` I think.  And that saves generating panic code so I think would be better.

Reply:
didn't realise that, oh well :upside_down_face: